src/shared/utilities/messages.ts

```diff
@@ -182,11 +183,114 @@ export async function showMessageWithCancel(
+ * It uniquely identifies a specific message.
+ */
+export type CancellableMessageId = string
+
+/**
+ * Provides support to create cancellable vscode message windows.
+ *
+ * To get started, you create a new message with a unique ID.
+ * Following that, you can easily update the message or cancel it.
+ *
+ * ---
+ *
+ * This class solves an issue of needing to manage duplicated
+ * messages, the interface simplifies the process.
+ */
+export class CancellableMessageManager {
+    static readonly timeoutMillis = 60000
+    private static messageMap: {
+        [msgId: CancellableMessageId]: { timeout: Timeout; progress: vscode.Progress<ProgressEntry> }
+    } = {}
+
+    /**
+     * Helper to generate unique ID from the given info
+     */
+    static idFromCwlInfo(
+        cwlGroup: CloudWatchLogsGroupInfo,
+        cwlParameters: CloudWatchLogsParameters
+    ): CancellableMessageId {
+        return `${getStringHash(JSON.stringify({ cwlGroup, cwlParameters }))}`
+    }
+
+    /**
+     * Starts a message if it does not exist, then applies the update with the given
+     * progress data.
+     */
+    static async updateOrStartMessage(
+        msgId: CancellableMessageId,
+        messageText: string,
+        progressEntry?: ProgressEntry,
+        timeoutMillis?: number
+    ): Promise<void> {
+        let message = this.messageMap[msgId]
+
+        if (message === undefined) {
+            await this.startMessage(msgId, messageText, timeoutMillis)
+            message = this.messageMap[msgId]
+        }
+
+        if (progressEntry !== undefined) {
+            message.progress.report(progressEntry)
+        }
+    }
+
+    /**
+     * @param timeoutMillis The amount of milliseconds till the message cancels itself.
+     */
+    static async startMessage(msgId: CancellableMessageId, messageText: string, timeoutMillis?: number): Promise<void> {
+        const message = this.messageMap[msgId]
+
+        if (message !== undefined) {
+            this.cancelMessage(msgId)
+        }
+
+        if (timeoutMillis === undefined) {
+            timeoutMillis = this.timeoutMillis
+        }
+
+        const timeout = new Timeout(timeoutMillis)
+        this.messageMap[msgId] = { timeout, progress: await showMessageWithCancel(messageText, timeout) }
+    }
+
+    static updateMessage(msgId: CancellableMessageId, progressUpdate: ProgressEntry): void {
+        const message = this.messageMap[msgId]
+
+        if (message === undefined) {
+            throw Error(`Cannot update progress for '${msgId}', it does not exist.`)
+        }
+
+        message.progress.report(progressUpdate)
+    }
+
+    static cancelMessage(msgId: CancellableMessageId): void {
+        const message = this.messageMap[msgId]
+
+        if (message === undefined) {
+            return
+        }
+
+        message.timeout.cancel()
+        delete this.messageMap[msgId]
+    }
+
+    static getTimeout(msgId: CancellableMessageId): Timeout {
```

Comment:
This is a leaky abstraction. A caller can cancel the timer (and therefore the message) without calling `cancelMessage`.

Reply:
Nice catch. I've created a new class `CancellableMessageTimeout` that extends `Timeout`. It will cancel the underlying `Timeout`, but then perform the necessary cleanup that `cancelMessage()` does.